### ISSUE TITLE
String comparison fails when using const and compiled Models.

### ISSUE THREAD

**Issue by DxsSucuk** [dotnet/efcore#32617]
### Description
When comparing an entity value with a const field value while also using compiled Models, EF Core creates a different SQL Statement and causes the query to fail. When you instead use a static field, it creates the correct SQL Statement and the query succeeds.

### Include your code
Here is a simple snippet with comments explaining the Issue.
```C#
    private static string name = "RĒD";
    private const string name2 = "RĒD";
    public static void Main(string[] args)
    {
        using (var context = new ApplicationContextDB())
        {
            // Use the working static field.
            if (!context.Tests.Any(c => c.Name == name))
            {
                // Using the const field as value, doesn't make a difference which one is being used to set.
                context.Add(new Test { Name = name2 });
                context.SaveChanges();
            }

            // The static test should work always.
            Console.WriteLine($"Static Test: {context.Tests.Any(c => c.Name == name)}");

            // The const test should fail when using Compiled Models, because the SQL statement build by EF Core is different.
            Console.WriteLine($"Const Test: {context.Tests.Any(c => c.Name == name2)}");

            // A workaround for this Issue is to outcommend the useModel call inside the ApplicationContextDB class.
        }
    }
```
I will be including a zip files containing code that can be used to reproduce it.
[EFCoreTest.zip](https://github.com/dotnet/efcore/files/13676283/EFCoreTest.zip)

### Include stack traces
No stacktrace.

### Include verbose output
No related to the Package Manager or CLI.

### Include provider and version information

EF Core version: 8.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 8.0
Operating system: Windows 11
IDE: Visual Studio 2022 17.8.3

**Comment by DxsSucuk** [comment]
This Issue does not happen on older EF-Core version.

**Comment by ajcvickers** [comment]
Note for triage: Regression in the Unicodeness of the const string in the query is not being generated correctly when the compiled model is being used.

With the compiled mode--see the non-Unicode string literal:

```sql
SELECT CASE
    WHEN EXISTS (
        SELECT 1
        FROM [Tests] AS [t]
        WHERE [t].[Name] = 'RED') THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END
```

Without the compiled model--the literal is Unicode:

```sql
SELECT CASE
    WHEN EXISTS (
        SELECT 1
        FROM [Tests] AS [t]
        WHERE [t].[Name] = N'RED') THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END
```

In EF7, we generate this with and without the compiled model:

```sql
SELECT CASE
    WHEN EXISTS (
        SELECT 1
        FROM [Tests] AS [t]
        WHERE [t].[Name] = N'RED') THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END
```

**Comment by DxsSucuk** [comment]
Still does not work in 8.0.1

**Comment by roji** [comment]
@DxsSucuk the fixed was merged for 8.0.2, which should be out in a few weeks.
